UI scale

TheCycoONE opened the pull request:
*Fixes #1210*

* Adds the option to scale to 200% or 300% for UI elements.
* Fixes the position of the English tooltip for the language dropdown entries
* Fixes the internal size of the confirm dialogs to match the visual size.
* Draws true type fonts larger rather than scaling for improved appearance.
* Nearest neighbour scaling for bitmap fonts to preserve readability.
* Dynamically scalable (all dialogs)
* Improves the appearance of bitmap text on the lose movie headline

What it looks like:

<details>
  <summary>Before. Default 100% scale. Press here 👈</summary>
  
<img width="1281" height="989" alt="Screenshot 2026-03-19 at 11 39 56" src="https://github.com/user-attachments/assets/3881f9c5-3ce2-4b38-a97f-214befbf4cc9" />
  
</details>

<details>
  <summary>After. 200% scale. Press here 👈</summary>
  
  <img width="1281" height="989" alt="Screenshot 2026-03-19 at 11 29 13" src="https://github.com/user-attachments/assets/cb50a3cb-f99f-4dda-9bb7-2aba26518abb" />
  
</details>

TheCycoONE committed:
Allow specifying a UI scale factor

Currently this is quite blurry, and the pick hospital screen is broken.

TheCycoONE committed:
Keep fonts crisp with UI scaling

TheCycoONE committed:
Add settings window for setting the scale

TheCycoONE committed:
Fix the cursor position when not drawn

When the cursor isn't drawn, e.g. the choose theme hospital directory
screen, it was in completely the wrong position. This fixes that by
compensating the cursor position.

There are still some gaps in game_ui, but the main menu and settings
screens are working well.

TheCycoONE committed:
Revert scaling full UI in favour of targetting scaling.

Scale panel sprites because that can't be helped.

Scale UIResizable, buttons, and panels

Scale some text drawing - though I will likely replace that with a draw argument later.

TheCycoONE committed:
Add scale factor to fonts

Use scale factor instead of canvas scale for window label text.

The visual difference with TTF fonts is very clear. Yey HiDPI.

TheCycoONE committed:
Reset option menu to apply scale change

TheCycoONE committed:
Apply font scale to confirm dialog

TheCycoONE committed:
UI scaling for tip of the day

TheCycoONE committed:
Progress towards bottom panel scaling

TheCycoONE committed:
Add option to draw scaled sprites

Use that instead of canvas:scale for panel sprites

TheCycoONE committed:
More bottom panel scaling

TheCycoONE committed:
scale the information window

TheCycoONE committed:
scale the jukebox

TheCycoONE committed:
Use new draw sprites API for map editor

TheCycoONE committed:
Scale pixel coordinates for panels,buttons,windows

Instead of storing the scaled height and width as state store whether
the Window / Panel is using scale independent pixels. If so perform the
adjustment in the draw/hitTest functions.

The goal is to avoid mixing scale information into the save state in
ways that will be hard to deal with.

Some regression here, mainly tooltips are not functioning

TheCycoONE committed:
Fix dragging

TheCycoONE committed:
Window positions also use scale coordinates

TheCycoONE committed:
Fix confirm dialog

TheCycoONE committed:
Fix information dialog

TheCycoONE committed:
Fix bottom panel

TheCycoONE committed:
scale watch and fix information

Improve hit tests / hover

TheCycoONE committed:
Start of save game compatibility

TheCycoONE committed:
Draw tooltips correctly for scale

TheCycoONE committed:
Add scaling for file/directory dialogs

TheCycoONE committed:
tooltip load fix

TheCycoONE committed:
Patient dialog

Include fix for line width

TheCycoONE committed:
Scale top menu

TheCycoONE committed:
Fix jukebox scale

TheCycoONE committed:
Add broken drag handling from new_game

Was redundant to UIResizable code.

TheCycoONE committed:
ui scalling for scrollbars and menu list dialogs

TheCycoONE committed:
information fix, do not shadow s

TheCycoONE committed:
build_room dialog scale

TheCycoONE committed:
scale edit room and place objects

TheCycoONE committed:
scale advisor

TheCycoONE committed:
Add ui_scale to busted mock

TheCycoONE committed:
Fix bottom panel hit test

TheCycoONE committed:
Fix windows config template for ui_scale

TheCycoONE committed:
some comments for parameter changes

TheCycoONE committed:
scale staff profile page

Add the ability to specify flags when loading a bitmap. This is needed
because linear scaling causes artifacts at the border where the face is
composed of multiple images.

TheCycoONE committed:
Add scale to staff dialog

Alberth289346 commented on CorsixTH/Src/th_lua_gfx.cpp:
```diff
@@ -43,6 +43,35 @@ SOFTWARE.
 #include FT_TYPES_H
 
 namespace {
+uint32_t flags_from_table(lua_State* L, int index) {
```
Are you aware that the draw flags also exist as named constants in Lua?

```
Lua/utility.lua
217:DrawFlags = {}
218:DrawFlags.FlipHorizontal  = 2^0
219:DrawFlags.FlipVertical    = 2^1
220:DrawFlags.Alpha50         = 2^2
221:DrawFlags.Alpha75         = 2^3
222:DrawFlags.AltPalette      = 2^4
223:DrawFlags.EarlyList       = 2^10
224:DrawFlags.ListBottom      = 2^11
225:DrawFlags.BoundBoxHitTest = 2^12
226:DrawFlags.Crop            = 2^13
```

`utility` also has a number of bit operation functions (`flag_*` functions around lines 190-210, and two more around lines 295-310)

TheCycoONE committed:
Use DrawFlags instead of support for flag names in binding

TheCycoONE committed:
update handling for scaled dialogs

Fonts and graphics need to be reloaded when the flags
changed.

TheCycoONE committed:
Fix missing variable for afterLoad place_objects

TheCycoONE committed:
ui_scale town map

TheCycoONE committed:
ui_scale fax

TheCycoONE committed:
ui_scale annual reports

TheCycoONE committed:
ui_scale bank_manager

TheCycoONE committed:
ui_scale drug_casebook

TheCycoONE committed:
ui_scale graphs

TheCycoONE committed:
ui_scale for hospital policy

TheCycoONE committed:
ui_scale progress report

TheCycoONE committed:
fix hospital_policy crash on scale with slider

TheCycoONE committed:
ui_scale graphs - fix text y position

TheCycoONE committed:
ui_scale research policy

TheCycoONE committed:
ui_scale staff management

TheCycoONE committed:
ui_scale machine dialog

TheCycoONE committed:
ui_scale machine menu

TheCycoONE committed:
ui_scale furnish corridor

TheCycoONE committed:
ui_scale queue

TheCycoONE committed:
ui_scale update

TheCycoONE committed:
ui_scale staff_raise

TheCycoONE committed:
wip: ui_scale map_editor

TheCycoONE committed:
ui_scale map_editor

TheCycoONE committed:
ui_scale: Fix size of language names

TheCycoONE committed:
ui_scale fix language tooltip position

TheCycoONE committed:
Fix size of confirm window

Without this fix the confirm window is fixed size even though visually
it adjusted to the size of the text which caused some strange
behaviours. Most visually it was unable to be dragged to the bottom of
the screen.

TheCycoONE committed:
ui_scale fix menu crash

TheCycoONE committed:
Fix several afterLoad methods

TheCycoONE committed:
ui_scale: fix potential crash on information dialog

TheCycoONE committed:
ui_scale Fix load issue on staff_management

tobylane reviewed (approved):
Minor points for the future

- Handle unexpected scale numbers
- Provide a way out of menu dialogs when the cancel button is out of view, eg esc.
- A cancel button/timeout on changing the scale factor? Maybe not needed, as the settings button and its dialog are in the centre of the screen.

It feels so authentic to have a larger ui!

TheCycoONE commented:
@lewri also I suppose you didn't test it. Full screen does render full screen again with those scales and those resolutions.

TheCycoONE committed:
Only include the machine button if it fits for scale

Also remove or add the machine button as the window resolution changes.

TheCycoONE commented:
<img width="2180" height="1756" alt="image" src="https://github.com/user-attachments/assets/e9409701-9b89-4559-93e0-68849d4c0f70" />

Fullscreen research window at 1920x1440

ARGAMX reviewed (approved):
Great work and a great improvement! 💪
I'm impressed 👍
We really missed that functionality.

Related to situations when the cancel button is out of view I found that I can drag any window up and down and this make cancel button visible, when I drag it DOWN (if I need to get button at bottom).

<img width="912" height="740" alt="Screenshot 2025-12-23 at 20 29 41" src="https://github.com/user-attachments/assets/ff989ffe-fa4e-479e-9a54-f01c22dc9557" />

I also noticed that some of the overlays now have borders that they didn't have at 100% before.

For example, you can notice how the advisor on 200% now has a thin, fashionable mustache.

<img width="342" height="237" alt="Screenshot 2025-12-23 at 20 11 40" src="https://github.com/user-attachments/assets/bb29b6c1-1e1b-48ab-a0d5-ad8f9187729a" />

I also noticed that at the 200% the elements on the salary increase window are misaligned.

<img width="237" height="342" alt="Screenshot 2025-12-23 at 20 11 40" src="https://github.com/user-attachments/assets/274c5760-99a1-4b25-b821-720e33d429ab" />

Also game freezes/crash when I try to drag that windows in edge positions.

Steps to reproduce:
1. Select 200% and 1280x960 and close the game.
2. Start the game.
3. Load attached save game. [0.69.1-Battenberg.sav.zip](https://github.com/user-attachments/files/24316274/0.69.1-Battenberg.sav.zip)
4. Open notification.
5. Drag the window.

Fractional scaling also came to mind and possible will be useful feature for this functionality in future.

Everything else is very cool! 🆒

Thank you, Cyco

TheCycoONE committed:
ui_scale fix position of markers in staff_rise dialog

TheCycoONE commented:
@flackr any ideas here? I seem to remember there were a few of these gap issues when direct_zoom was first added. I think I'd prefer if the animation wasn't split into different textures.

TheCycoONE committed:
ui_scale fix build room preview size

Was an animation not a sprite so it does not support scaleFactor.

TheCycoONE committed:
ui_scale: fix click area for UIStaff

Should iterate when right clicking on the name.
